src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,777 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "iotjs_def.h"
+#include "iotjs_module_buffer.h"
+#include "iotjs_module_mqtt.h"
+
+
+#include "iotjs_handlewrap.h"
+#include "iotjs_reqwrap.h"
+
+static void iotjs_mqtt_client_destroy(iotjs_mqtt_client_t *mqtt_client) {
+  IOTJS_RELEASE(mqtt_client);
+}
+
+static const jerry_object_native_info_t mqtt_client_native_info = {
+  .free_cb = (jerry_object_native_free_callback_t)iotjs_mqtt_client_destroy
+};
+
+static iotjs_mqtt_client_t *iotjs_mqtt_client_create(
+    const jerry_value_t jobject) {
+  iotjs_mqtt_client_t *mqtt_client = IOTJS_ALLOC(iotjs_mqtt_client_t);
+
+  jerry_set_object_native_pointer(jobject, mqtt_client,
+                                  &mqtt_client_native_info);
+
+  return mqtt_client;
+}
+
+
+IOTJS_DEFINE_NATIVE_HANDLE_INFO_THIS_MODULE(mqtt_client);
+
+
+static jerry_value_t mqtt_client_connack_error(const unsigned char error_code) {
+  switch (error_code) {
+    case UNACCEPTABLE_PROTOCOL:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: unacceptable protocol");
+    case BAD_IDENTIFIER:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: bad client identifier");
+    case SERVER_UNAVIABLE:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: server unaviable");
+    case BAD_CREDENTIALS:
+      return JS_CREATE_ERROR(
+          COMMON, "MQTT: Connection refused: bad username or password");
+    case UNAUTHORISED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Connection refused: unauthorised");
+    default:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+/*
+ * Waits for a uint32_t remaining_length.
+ * Returns the encoded value with the correct size.
+ */
+static size_t iotjs_encode_remaining_length(uint32_t len,
+                                            unsigned char *buffer) {
+  size_t rc = 0;
+  do {
+    unsigned char d = len & 0x7F;
+    len >>= 7;
+    if (len > 0) {
+      d |= 0x80;
+    }
+    buffer[rc++] = d;
+  } while (len > 0);
+
+  return rc;
+}
+
+
+static size_t get_remaining_length_size(uint32_t len) {
+  uint8_t n = 0;
+  while (len != 0) {
+    len >>= 7;
+    n++;
+  }
+
+  return n;
+}
+
+
+static uint32_t iotjs_decode_remaining_length(char *buffer, size_t *offset) {
+  unsigned char c;
+  uint32_t remaining_length = 0;
+  uint32_t length = 0;
+  uint32_t multiplier = 1;
+
+  do {
+    if (++length > IOTJS_MODULE_MQTT_MAX_REMAINING_LENGTH_BYTES) {
+      return JS_CREATE_ERROR(COMMON, "MQTT: The arrived message is too big");
```

Comment:
You should return with a special value to show error. E.g. `UINT32_MAX`, since the message size is limited to 256 Mbyte.